test_plans/atomic_ref.asciidoc

```diff
@@ -0,0 +1,480 @@
+
+==== Common constructors
+
+Create an `atomic_ref` object using `atomic_ref(T&)` and an other object using copy constructor in device code using the following values as `DefaultOrder` template parameter:
+
+* `memory_order::relaxed`
+* `memory_order::acq_rel`
+* `memory_order::seq_cst`
+
+The following values as `DefaultScope` template parameter:
+
+* `memory_scope::work_item`
+* `memory_scope::sub_group`
+* `memory_scope::work_group`
+* `memory_scope::device`
+* `memory_scope::system`
+
+And the following values as `Space` template parameter:
+
+* `access::address_space::global_space`
+* `access::address_space::local_space`
+* `access::address_space::generic_space`
+
+After calling constructors check for:
+
+* `atomic_ref(T&)`
+
+** If the stored value match with the value passed as `T&` using
+`T load(
+    memory_order order = default_read_order
+    memory_scope scope = default_scope) const`.
+Check returned value type of `load` member function.
+
+* `atomic_ref(const atomic_ref&) noexcept`
+
+** If the members of the newly created object match that of `const atomic_ref&` and the referenced value is valid using
+`T load(
+    memory_order order = default_read_order
+    memory_scope scope = default_scope) const`.
+
+==== Member functions
+
+To test member functions create an `atomic_ref` object using `atomic_ref(T&)` in device code using the following values as `DefaultOrder` template parameter:
+
+* `memory_order::relaxed`
+* `memory_order::acq_rel`
+* `memory_order::seq_cst`
+
+And the following values as `DefaultScope` template parameter:
+
+* `memory_scope::work_item`
+* `memory_scope::sub_group`
+* `memory_scope::work_group`
+* `memory_scope::device`
+* `memory_scope::system`
+
+And the following values as `Space` template parameter:
+
+* `access::address_space::global_space`
+* `access::address_space::local_space`
+* `access::address_space::generic_space`
+
+Note: for the member functions tests below that take `memory_order` and `memory_scope` as arguments, the following values of these parameters are used if the `FULL CONFORMANCE` mode is enabled:
+
+* For `memory_order`:
+
+** `memory_order::relaxed`
+** `memory_order::acquire`
+** `memory_order::release`
+** `memory_order::acq_rel`
+** `memory_order::seq_cst`
+
+* For `memory_scope`:
+
+** `memory_scope::work_item`
+** `memory_scope::sub_group`
+** `memory_scope::work_group`
+** `memory_scope::device`
+** `memory_scope::system`
+
+It should be noted that `memory_order::acquire` is used only for read operations, `memory_order::release` for write operations and `memory_order::acq_rel` for read-modify-write operations.
+
+If the `FULL CONFORMANCE` mode is disabled only default values for functions arguments is used which are determined by `atomic_ref` template parameters that which object was created with.
+
+===== Common member functions
+
+`bool is_lock_free()`
+
+Check if the function exists and returns `atomic_ref::is_always_lock_free`.
```

Comment:
This doesn't look correct: if `atomic_ref::is_always_lock_free` is `false`, it doesn't mean that this particular instance of `atomic_ref` can't perform lock-free operations.

It seems to me that the check we want to see is that if `atomic_ref::is_always_lock_free` is `true`, then `is_lock_free()` must also return `true`. If `atomic_ref::is_always_lock_free` is `false`, then we can't make any assumptions about result of `is_lock_free()`.